### DISCUSSION TITLE
Enhancements

### DISCUSSION THREAD

**Discussion by thomasj02** [VsonicV/es-fine-tuning-paper#18] [General]
So I tried out a few of the enhancements that the paper mentioned:
* Adam instead of vanilla SGD
* Fitness shaping
* Mirrored sampling

What I found:
* Normal Adam was pretty terrible for reasons I don't understand. The gradients rapidly collapsed to zero and all of the samples had a fitness of 0.60. 
* Fitness shaping didn't seem to help, and I notice the original implementation already has some fitness shaping that Z-scores the fitnesses
* Mirrored sampling got me to a max score of 0.25 on Qwen-Instruct-0.5B. The paper only reports accuracy for 0.5B rather than the reward, so I don't know how this compares to the reward with the non-mirrored sampling approach.

Also a major caveat: I only ran a single run for each of these tests, so it's very possible that some of the changes didn't work just due to luck. Also I haven't tried combining any of the enhancements (e.g. Adam + Mirrored sampling)

**Comment by conorfhayes** [comment]
Interesting, thanks for looking into this. Mirror sampling may be effected by the population size also. How are the implementing mirror sampling? You should try AdamW in your setup. It would be interesting if you could share your code.

We are planning to add and test these enhancements eventually.

**Reply by VsonicV** [reply]
I assume @thomasj02 is using the same setup as ours, i.e., the population size is 30. @conorfhayes Why do you think mirrored sampling will be affected by the population size? In my opinion, mirrored sampling is just another way to sample the perturbations, so it should further improve the sampling efficiency regardless of the population size.

**Reply by conorfhayes** [reply]
Yes, the same setup, but can you share the implementation of Adam and other additions? It would be interesting to see.

**Comment by VsonicV** [comment]
@thomasj02 Thanks for the testing. The mirrored sampling seems to work in you case. The training reward in the countdown task gives 0.1 score to a wrong answer with correct formatting, so the corresponding accuracy may be slightly lower than the reward. I assume the 0.25 here is your training reward after 500 iterations? if so, you can test the accuracy of the fine-tuned model using the samples in `countdown/data/countdown.json` (do not use the first 200 training samples).

**Comment by thomasj02** [comment]
@VsonicV 
Apologies for the delayed response, I've been in the process of looking for new job positions
That's correct that 0.25 is the training reward, but it's after about 200 iterations. I've discarded the run data but I can re-run it and check the out of sample accuracy. 

You can also see some training data at https://wandb.ai/thomasj02/es-fine-tuning-countdown/workspace
I ran them all for a relatively short number of iterations because I'm only using my personal workstation with a single 5090 GPU. 

@conorfhayes - @VsonicV is correct, I kept the population size of 30 and just did simple mirrored sampling where I generated half as many perturbations and multiplied each one by -1 and +1

**Comment by Green0-0** [comment]
In https://github.com/VsonicV/es-fine-tuning-paper/discussions/22 I had claimed that mirror sampling did not work, but I have made an update to my code as it turns out I implemented mirrored sampling wrong (I kept the normalization of (r - r_mean) / std, but this should've been removed). The new graphs are on my wandb project page.

Admittedly though the performance didn't increase by that much. I think with a bit more LR tuning, I can get the mirrored run to beat the normal run, but that wouldn't be a valid result without a hyperparameter sweep.

Also, while I've never gotten momentum to beat the default update rule (haven't tried adam), I did not see the gradients collapsed to zero; if anything it seems that with the default (too high) momentum coefficient the gradients become too large and the model diverges.
